### ISSUE TITLE
Drag and Drop for changing parent-child relationship

### ISSUE THREAD

**Issue by Chethan-7** [bumbeishvili/org-chart#155]
Hi,

Does the current plugin support drag-and-drop feature where i could drag a node and drop it on another node and dragged node becomes the child?
I am looking for some pointers if this is already explored.

Thanks,
Chethan

**Comment by bumbeishvili** [comment]
No, it does not currently support it, but it’s planned for the future

**Comment by surajbabar** [comment]
Any update on when this will be available?
If it is already can you share more details on how to enable it ?

**Comment by bumbeishvili** [comment]
I don't expect it to be implemented in the near future (at least for the next 4-6 months)

**Comment by A7medMo7ammed20** [comment]
Any Updates?

**Comment by bumbeishvili** [comment]
@A7medMo7ammed20  there is a potential partnership opportunity for paying for the implementation of drag-drop functionality.

So there is a chance that it might get implemented within the next 3 months

**Comment by joelcoxokc** [comment]
@bumbeishvili I am able to contribute to this effort. 
We need this feature ASAP.

**Comment by joelcoxokc** [comment]
@A7medMo7ammed20 

Until they get this feature in place.. 

I am doing this.. 

You still have to handle hiding the children and the edges. But this is the beginning of getting a node to drag

I do this right after I call chart.render()

```ts

let dragStartX: number;
let dragStartY: number;
let newParent: GenNodeData | null;
const nodesSelection = state['nodesWrapper']
.selectAll('g.node')
.on('mouseover', (d: any, n: any) => {
    console.log({ d, n });
    newParent = n;
})
.on('mouseout', (d: any, n: any) => {
    console.log({ d, n });
    newParent = null;
})
.call(
    d3
        .drag()
        .on('start', (d) => {
            const width = d.subject.width;
            const half = width / 2;
            const x = d.x - half;
            dragStartX = x;
            dragStartY = parseFloat(d.y);
        })
        .on('drag', function (d) {
            const g = d3.select(this);
            dragStartX += parseFloat(d.dx);
            dragStartY += parseFloat(d.dy);
            g.attr('transform', 'translate(' + dragStartX + ',' + dragStartY + ')');
        })
        .on('end', (d) => {
            console.log({ newParent });
        })
);
```

**Comment by bumbeishvili** [comment]
There is a #307 PR if you want to have a look and make it work. 

The work I plan to do with Drag & drop is not quick and it's quite substantial, so won't merge or add anything related to this  meantime

**Comment by joelcoxokc** [comment]
Here is my drag example.. I copied what I could from our production app.
https://stackblitz.com/edit/js-bmyjfc?file=index.html

This has undo/redo/cancel built into it.
Also each node can be draggable or droppable.

I made a button to start the drag process.. otherwise there are all kinds of issues when trying to pan while dragging is enabled.
So you have to click organize to start the drag-drop.

@bumbeishvili Feel free to fork this and add it to the examples

**Comment by bumbeishvili** [comment]
Nice, I'd implement it differently but good to have it implemented anyway

**Comment by joelcoxokc** [comment]
> Nice, I'd implement it differently but good to have it implemented anyway

What would you do differently I would love to know as we are using this in production

**Comment by bumbeishvili** [comment]
First I am very much impressed that you implemented drag & drop without modifying source code (and I am also proud too because it's possible)

• I'd try to overcome the need to have a drag functionality dependent on the button (and it's not an easy )
• I'd build dragging functionality inside the chart and I'd expose some methods for controlling how you want the behavior of the org chart to be, specifically I'd have following methods exposed
```
onDragStart,   // Styling dragging node
onDrag, 
onDragEnd,  // Unstiling dragging node
onTarget(({fromData,toData,fromId,toId,fromNode,toNode})),   // Styling target node 
outTarget(({fromData,toData,fromId,toId,fromNode,toNode})),  // Unstyling target node
onNodeDrop({idFrom,idTo,nodes}=>{}) // changing the data and rendering the graph
```
• I'd make it possible to drop over the node once it touches another node

That said, again, your implementation is quite good since it is done without modifying anything in the core codebase and in theory will not interfere with future updates and in a way it's safe solution

**Comment by joelcoxokc** [comment]
> First I am very much impressed that you implemented drag & drop without modifying source code (and I am also proud too because it's possible)
> 
> • I'd try to overcome the need to have a drag functionality dependent on the button (and it's not an easy ) • I'd build dragging functionality inside the chart and I'd expose some methods for controlling how you want the behavior of the org chart to be, specifically I'd have following methods exposed
> 
> ```
> onDragStart,   // Styling dragging node
> onDrag, 
> onDragEnd,  // Unstiling dragging node
> onTarget(({fromData,toData,fromId,toId,fromNode,toNode})),   // Styling target node 
> outTarget(({fromData,toData,fromId,toId,fromNode,toNode})),  // Unstyling target node
> onNodeDrop({idFrom,idTo,nodes}=>{}) // changing the data and rendering the graph
> ```
> 
> • I'd make it possible to drop over the node once it touches another node
> 
> That said, again, your implementation is quite good since it is done without modifying anything in the core codebase and in theory will not interfere with future updates and in a way it's safe solution

I need to implement the same thing in another interface in our app.. 
Depending on complexity, I will try to implement this in the core library and send you a PR

Regarding when the node touches another node... 
I tried to achieve this best I could, but the dragging node could be touching between 1 and three nodes, based on node margin. 
So I implemented to when the dragging node is more that 50% over the drop on both x & y, I highlight the node

I could probably do (50% - nodeMargin)

**Comment by joelcoxokc** [comment]
@bumbeishvili https://github.com/bumbeishvili/org-chart/pull/350
